File: src/graphics/Cam.java
package graphics;

import managers.GameManager;
import math.components.*;
import math.util.UtilVector;
import static math.util.UtilVector.createVector;
import math.util.Utilities;

public class Cam {
    double fov, rotSpeed;
    public Point puntoDeLaCamara, position;
    GameManager gm;
    public Vector vectorNormal, vectorALaCam, vector_X_local, vector_Y_local;
    private Plane plano_camara;
    public Rect axisX, axisY; //Ejes en 2d de la screen
    public Point pivotX, pivotY; //Puntos para poder rotar correctamente los ejes al rotar la cam
    public Cam(int field_of_view){
        this.fov = field_of_view;
        this.rotSpeed = 1;
        this.gm = GameManager.getInstance();
        this.position = gm.player.position;
        this.puntoDeLaCamara = new Point(position.x+fov, position.y, position.z);
        this.vectorALaCam= createVector(puntoDeLaCamara, position);
        setUpAxis();
        actualizar_plano();
    }

    public void update(){
        actualizar_plano();
        info3Dto2D();
    }

    private void setUpAxis(){
        //Para poder crear correctamente los ejes el ángulo vertical (El ángulo del vectorALaCam
        //con el eje y) tiene que ser 90
        double vertAng = Math.toDegrees(UtilVector.getAngle(new Vector(0, 1, 0), vectorALaCam.normalize()));
        if (vertAng != 90) this.puntoDeLaCamara = new Point(position.x+fov, position.y, position.z);
        //Como el ángulo vertical es 90, el vector entre el p cam y el p cam + j es el eje y en 2d
        this.pivotY = new Point(puntoDeLaCamara.x, puntoDeLaCamara.y+1, puntoDeLaCamara.z);
        Vector vectorY = createVector(pivotY, puntoDeLaCamara);
        this.axisY = new Rect(puntoDeLaCamara, vectorY);
        //El eje x 2d es perpendicular tanto al eje y 2d, como al vector a la cam por lo que
        //el p vectorial entre el vectorALaCam y el eje y 2d es el eje x 2d
        Vector vectorX = UtilVector.producto_vectorial(vectorY, vectorALaCam.normalize());
        this.axisX = new Rect(puntoDeLaCamara, vectorX);
        this.pivotX = puntoDeLaCamara.add(vectorX);
    }

    public void rotarCamara(float horizontal, float vertical){
        Point puntoCamaraCentrado = this.puntoDeLaCamara.subtract(position);
        Point pivotXCentrado = this.pivotX.subtract(position);
        Point pivotYCentrado = this.pivotY.subtract(position);
        if (horizontal != 0){
            puntoCamaraCentrado.rotar(0,horizontal,0);
            pivotXCentrado.rotar(0,horizontal,0);
            pivotYCentrado.rotar(0,horizontal,0);
        }
        if (vertical != 0){
            puntoCamaraCentrado.rotateVertically(vertical);
            pivotXCentrado.rotateVertically(vertical);
            pivotYCentrado.rotateVertically(vertical);
        }
        this.puntoDeLaCamara = puntoCamaraCentrado.add(position);
        this.pivotX = pivotXCentrado.add(position);
        this.pivotY = pivotYCentrado.add(position);
        this.axisX = new Rect(puntoDeLaCamara, pivotX);
        this.axisY = new Rect(puntoDeLaCamara, pivotY);
        vector_X_local.normalize();
        this.actualizar_plano();
    }

    private void actualizar_plano(){
        this.position = gm.player.position;
        this.vectorNormal = createVector(puntoDeLaCamara, position);
        this.plano_camara = new Plane(puntoDeLaCamara, vectorNormal);
        this.vectorALaCam = createVector(puntoDeLaCamara, position);
        this.vector_X_local = UtilVector.producto_vectorial(vectorALaCam, UtilVector.getDown());
        this.vector_Y_local = UtilVector.producto_vectorial(vectorALaCam, vector_X_local);
    }

    private void info3Dto2D(){
        gm.info2d.clear();
        for (Point i : gm.info3d){
            Point2 p = watch(i);
            if (p != null){
                gm.info2d.add(p);
            }
        }
    }

    private Point2 watch(Point point){
        Rect ray = new Rect(position, point);
        Point cutPoint = Utilities.interseccion_recta_plano(ray, plano_camara);
        if (cutPoint == null){
            return null;
        }
        double cutT = ray.getT(cutPoint);
        if (!(cutT > 0 && cutT<ray.getT(point))) {
            return null;
        }

        Vector vectorAlPunto = createVector(cutPoint, puntoDeLaCamara);
        double distancia = vectorAlPunto.getMod();

        double[] resultado = UtilVector.getCuadrante_y_angulo(vectorAlPunto,vector_X_local,vector_Y_local);
        double signx = resultado[0];
        double signy = resultado[1];
        double angulo = resultado[2];
        double x_local = distancia*Math.cos(angulo);
        double y_local = distancia*Math.sin(angulo)*signy;
        return new Point2((float)x_local, (float) y_local);
    }

    private Point2 watch2(Point target){
        Rect ray = new Rect(position, target);
        Point cutPoint = Utilities.interseccion_recta_plano(ray, plano_camara);
        if (cutPoint == null) return null;

        //Si la distancia entre la cam y cutPoint es mayor que la distancia entre la cam y el objetivo
        //Quiere decir que el objetivo no se encuentra en nuestro campo de visión
        Vector cam_cp = createVector(cutPoint, position);
        Vector cam_tg = createVector(target, position);
        if (cam_cp.getMod() > cam_tg.getMod()) return null;

        //Si la distancia entre la cam y el target es menor que la distancia entre el punto de la cam
        //y el target quiere decir que el objetivo no se encuentra en nuestro campo de visión
        Vector pcam_tg = createVector(target, puntoDeLaCamara);
        if (cam_tg.getMod() < pcam_tg.getMod()) return null;



        return null;
    }

    public double getRotSpeed() {
        return rotSpeed;
    }

    public void addPosition(Vector vector){
        puntoDeLaCamara.add(vector);
        pivotX.add(vector);
        pivotY.add(vector);
    }

    public void substractPosition(Vector vector){
        puntoDeLaCamara.subtract(vector);
        pivotX.subtract(vector);
        pivotY.subtract(vector);
    }

    //falta añadir comentarios de python y funciones
}

File: python/Cam.py
from Util import *
from GameManager import GameManager
from InputManager import Input

class Cam:
    def __init__(self, d, gm:GameManager):
        self.centro = gm.player.position
        self.d = d
        self.puntoDeLaCamara = Point(self.centro.x+self.d, self.centro.y, self.centro.z)
        self.gm = gm
        self.vrc = 2 #Velocidad de rotacion de la camara
        #Estoy asumiendo que cuando se inicializa por primera vez la cam el jugador mira hacia adelante
        # self.axisX2D = Rect(Point(self.puntoDeLaCamara.x, self.puntoDeLaCamara.y+1, self.puntoDeLaCamara.z), self.puntoDeLaCamara)
        # self.axisY2D = Rect(Point(self.puntoDeLaCamara.x, self.puntoDeLaCamara.y+1, self.puntoDeLaCamara.z).rotar(0, -90, 0), self.puntoDeLaCamara)
        self.actualizar_plano()
        self.info3Dto2D()

    def update(self):
        self.actualizar_plano()
        self.info3Dto2D()

    def rotarCamara(self, horizontal=0,vertical=0):
        puntoCamaraCentrado = self.puntoDeLaCamara - self.centro
        print(puntoCamaraCentrado)
        if horizontal is not 0:
            puntoCamaraCentrado.rotar(angulo_y=horizontal)
        if vertical is not 0:
            puntoCamaraCentrado.rotateVertically(vertical)
        print(puntoCamaraCentrado)
        self.puntoDeLaCamara = puntoCamaraCentrado + self.centro
        self.vector_X_local.normalize()    
        self.actualizar_plano() #Actualizar posición plano cámara


    def actualizar_plano(self):
        self.centro = self.gm.player.position
        
        # self.puntoDeLaCamara = Point(self.centro.x+self.d, self.centro.y, self.centro.z)
        
        self.vectorNormal= Util.Vector.createVector(self.puntoDeLaCamara,self.centro)
        self.planoCam = PlaneNor(self.vectorNormal,self.puntoDeLaCamara)
        #Conseguir el vector que se refiere al eje x local
        self.vectorALaCamara = Util.Vector.createVector(self.puntoDeLaCamara,self.centro)
        self.vector_X_local = Util.Vector.producto_vectorial(self.vectorALaCamara,Util.Vector.getDown())
        self.vector_Y_local = Util.Vector.producto_vectorial(self.vectorALaCamara,self.vector_X_local)

    # def watchMario(self, point:Point):
    #     ray = Rect(self.centro, point)
    #     cutPoint = Util.interseccion_recta_plano(ray, self.planoCam)
    #     #Buscar los puntos de corte con los ejes 2D para encontrar las coordenadas 2D
    #     pInX = Util.interseccion_recta_recta(Rect(cutPoint, self.axisY2D.vDir), self.axisX2D)
    #     pInY = Util.interseccion_recta_recta(Rect(cutPoint, self.axisX2D.vDir), self.axisY2D)
    #     #Coordenadas en 2d
    #     x2d = Util.Vector.createVector(cutPoint, pInX).getMod()
    #     y2d = Util.Vector.createVector(cutPoint, pInY).getMod()

    def info3Dto2D(self):
        for i in self.gm.info3d:
            p = self.watchDiego(i)
            # print(p)
            if p is not None:
                self.gm.info2d.add(p)
    
    def watchDiego(self, point:Point):
        # print("Punto",point.name)
        # print(self.centro)
        
        ray = Rect(self.centro,point)
        
        # print("vectoralpunto",ray.vDir)
        cutPoint = Util.interseccion_recta_plano(ray, self.planoCam)
        if cutPoint is None:    
            return None
        #el parametro T crece linealmente de manera positiva hacia el segundo punto encontes el parametro t para el plano debe estar entre el origen 0 y el punto origen
        if not (0 >= ray.getT(cutPoint) >= ray.getT(point)):
            return None
        #Conseguir el vector del centro al punto creo q se puede mejorar
        vectorAlPunto = Util.Vector.createVector(cutPoint,self.puntoDeLaCamara)
        #COnseguir la distancia
        distancia = vectorAlPunto.getMod()
        # print("xlocal:",self.vector_X_local,"punto corte:",cutPoint,"punto_camara:",self.puntoDeLaCamara)
        #Conseguir el angulo entre anmbos vector los devuelve en angulos
        
        # print("xlocal:",self.vector_X_local,"vectoralpunto:", vectorAlPunto)
        [signx,signy,angulo] = Util.Vector.getCuadrante_y_angulo(vectorAlPunto,self.vector_X_local,self.vector_Y_local)
        #print("angulo:",math.degrees(angulo), "distancia:",distancia)
        x_local = distancia*math.cos(angulo)
        y_local = distancia*math.sin(angulo)*signy
        # print("x:",x_local,"y:",y_local)
        return Point2D(x_local,y_local)
    
    
    def rotacioCamaraY(self,angY):
        vectorXenEje = self.vector_X_local
        ang_nor = Util.Vector.angulo_entre_vectores(Util.Vector.getLeft(),vectorXenEje)
        vectorXenEje.rotar(angulo_y=math.degrees(ang_nor))
        if (math.isclose(vectorXenEje.normalize().x,Util.Vector.getLeft().x)):
            self.puntoDeLaCamara.rotar(angulo_y=math.degrees(ang_nor))
            self.puntoDeLaCamara.rotar(angulo_z=math.degrees(angY))
            self.puntoDeLaCamara.rotar(angulo_y=-1*math.degrees(ang_nor))
            
        else:
            self.puntoDeLaCamara.rotar(angulo_y=-1*math.degrees(ang_nor))
            self.puntoDeLaCamara.rotar(angulo_z=math.degrees(angY))
            self.puntoDeLaCamara.rotar(angulo_y=math.degrees(ang_nor))
            
        self.actualizar_plano()
    
    def listenRotationCamera(self, input):
        if input == None:
            return
        elif input == Input.UP:
            
            self.rotarCamara(vertical=self.vrc)
        elif input == Input.DOWN:
            self.rotarCamara(vertical=self.vrc*-1)
        elif input == Input.LEFT:
            self.rotarCamara(horizontal=self.vrc)
        elif input == Input.RIGHT:
            self.rotarCamara(horizontal=self.vrc*-1)
        elif input == Input.LEFTUP:
            self.rotarCamara(vertical=self.vrc,horizontal=self.vrc)
            
        elif input == Input.LEFTDOWN:
            self.rotarCamara(vertical=self.vrc*-1,horizontal=self.vrc)

        elif input == Input.RIGHTUP:
            self.rotarCamara(vertical=self.vrc,horizontal=self.vrc*-1)

        elif input == Input.RIGHTDOWN:
            self.rotarCamara(vertical=self.vrc*-1,horizontal=self.vrc*-1)


